### ISSUE TITLE
Invalid json format

### ISSUE THREAD

**Issue by Alaizabel** [transistive/neo4j-php-client#2]
Hello Ghlen,

the authentication error is gone after the update of the package. It is replaced by a format error:

`Fatal error: Uncaught Laudis\Neo4j\Client\Exception\Neo4jException: Invalid json format in C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php:239 Stack trace: #0 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(267): Laudis\Neo4j\Client\HttpDriver\SessionApi4->throwIfErrorIsReceived(Object(GuzzleHttp\Psr7\Response)) #1 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(90): Laudis\Neo4j\Client\HttpDriver\SessionApi4->handleRequest(Object(GuzzleHttp\Psr7\Request)) #2 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\Pipeline.php(48): Laudis\Neo4j\Client\HttpDriver\SessionApi4->flush(Object(Laudis\Neo4j\Client\HttpDriver\Pipeline)) #3 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(50): Laudis\Neo4j\Client\HttpDriver\Pipeline->run() #4 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\Connection\Connection in C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php on line 239`

I also cleared the cache of the Chrome in total as suggested in the other issue by Occul77rick5 but that didnt solved the issue. Also a restart didnt helped here. 

Best regards,
Alai

**Comment by Occul77rick5** [comment]
Are you using` php artisan serve` to start the site or navigating the full path in the url?

**Comment by transistive** [comment]
Hello @Alaizabel,

The error your received means the HTTP response from the server is not a valid json format. Is there any way you can log the response? How are you hosting the neo4j server? I can try and simulate your environment.

Kind regards,

Ghlen

**Comment by Alaizabel** [comment]
Hi,

@Occul77rick5 
This is a standard XAMPP environment not a Laravel. The Apache is started via GUI (Control). 

This is what the APACHE access log has:
`
::1 - - [17/Sep/2020:19:27:49 +0200] "GET /playground%20CODE/test_db.php HTTP/1.1" 200 1168 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:80.0) Gecko/20100101 Firefox/80.0"
::1 - - [17/Sep/2020:19:27:52 +0200] "GET /favicon.ico HTTP/1.1" 200 30894 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:80.0) Gecko/20100101 Firefox/80.0"`

Here is the apache error log:

`[Thu Sep 17 19:30:34.802862 2020] [php7:error] [pid 18500:tid 1896] [client ::1:61401] PHP Fatal error:  Uncaught Laudis\\Neo4j\\Client\\Exception\\Neo4jException: Invalid json format in C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\HttpDriver\\SessionApi4.php:239\nStack trace:\n#0 C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\HttpDriver\\SessionApi4.php(267): Laudis\\Neo4j\\Client\\HttpDriver\\SessionApi4->throwIfErrorIsReceived(Object(GuzzleHttp\\Psr7\\Response))\n#1 C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\HttpDriver\\SessionApi4.php(90): Laudis\\Neo4j\\Client\\HttpDriver\\SessionApi4->handleRequest(Object(GuzzleHttp\\Psr7\\Request))\n#2 C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\HttpDriver\\Pipeline.php(48): Laudis\\Neo4j\\Client\\HttpDriver\\SessionApi4->flush(Object(Laudis\\Neo4j\\Client\\HttpDriver\\Pipeline))\n#3 C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\HttpDriver\\SessionApi4.php(50): Laudis\\Neo4j\\Client\\HttpDriver\\Pipeline->run()\n#4 C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\Connection\\Connection in C:\\xampp\\htdocs\\playground CODE\\vendor\\laudis\\neo4j-php-client\\src\\HttpDriver\\SessionApi4.php on line 239`

A print_r(apache_response_headers()); gives back 
`Array ( [X-Powered-By] => PHP/7.4.10 ) `

on the authentication request. For the second query request I do not get a response_header but only the error-code.

The Neo4J-server is standard downloaded from the webpage and saved on a local disk. It's been started via cmd with the classical /bin/neo4j console. 

`D:\Neo\neo4j-community-4.1.1\bin>neo4j console
2020-09-17 17:27:43.823+0000 INFO  Starting...
2020-09-17 17:27:46.661+0000 INFO  ======== Neo4j 4.1.1 ========
2020-09-17 17:27:48.777+0000 INFO  Performing postInitialization step for component 'security-users' with version 2 and status CURRENT
2020-09-17 17:27:48.777+0000 INFO  Updating the initial password in component 'security-users'
2020-09-17 17:27:51.481+0000 INFO  Bolt enabled on localhost:7687.
2020-09-17 17:27:52.416+0000 INFO  Remote interface available at http://localhost:7474/
2020-09-17 17:27:52.416+0000 INFO  Started.`

The PHP / APACHE is a XAMPP Installation (Version 7.4.10 -- CP build 3.2.4) with PHP 7.4.10.

The test-call is made on a standard chrome browser via URL call on the localhost. Anything else I can do / test to help you to simulate the environment?

Alai

**Comment by transistive** [comment]
@Alaizabel Thanks for the information. I was hoping to be able to get the HTTP response from the neo4j server. If this is not possible, can you also share your PHP code were you setup your client?

Also, can you log in to your database with the neo4j client application? I'm curiuous if that works.

Kind regards,

Ghlen

**Comment by Alaizabel** [comment]
Hi Ghlen,

if you explain me how to fetch that information I'll give it a try.

The PHP code for the connect is simple - it's the example from the graphaware client:

`  require_once 'vendor/autoload.php';
  use Laudis\Neo4j\Client\ClientBuilder;

$client = ClientBuilder::create()
    ->addConnection('default', 'http://neo4j:test123@localhost:7474') 
    ->build();

$client->run('MATCH (n) RETURN COUNT(n) AS number;');`

Direct access to the Neo4J DB via 7474 is possible:

You are connected as user neo4j
to neo4j://localhost:7687
Connection credentials are stored in your web browser.

Also the CURL (see other issue) still works and returns a valid result.

The desktop as well shows a working instance no errors: [https://ibb.co/w4LMjvT](url)
Also the tools like Bloom etc. work and show valid results.

I'd dare to say the neo4j server is fine.

Best,
Alai

**Comment by transistive** [comment]
Dear Alai,

I took a gamble while reading your setup. If your neo4j database is not hosted on the same machine, it is possible a CORS Preflight request will be sent. This in turn should mean a 204 response code will be returned before the actual payload request is made. Since there is no code to handle this type of request, a strict decoding of json would indeed crash, as an empty string is invalid json.

I hope this theory holds up in practice. You can update to version 5.0.2 now to try it out :)

If this does not work, our best bet is to capture the response of the neo4j server using a tool like wireshark. But let's not get ahead of ourselves :smile: 

Please respond in both a success or failure of the patch so I know if I can close or continue to pursue the issue.

Kind regards,
Ghlen

**Comment by Alaizabel** [comment]
Hello Ghlen,

no, all programs run on the same unit. The php script is called over localhost by URL. I am sorry but I am afraid this won't be the reason. I was so hoping you would be able to find and fix it - I feel bad to cause so much work on your side. 

Just to be sure I updated the package and re-run the script with this error:

`Fatal error: Uncaught Laudis\Neo4j\Client\Exception\Neo4jException: Invalid json format in C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php:245 Stack trace: #0 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(273): Laudis\Neo4j\Client\HttpDriver\SessionApi4->throwIfErrorIsReceived(Object(GuzzleHttp\Psr7\Response)) #1 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(92): Laudis\Neo4j\Client\HttpDriver\SessionApi4->handleRequest(Object(GuzzleHttp\Psr7\Request)) #2 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\Pipeline.php(48): Laudis\Neo4j\Client\HttpDriver\SessionApi4->flush(Object(Laudis\Neo4j\Client\HttpDriver\Pipeline)) #3 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(52): Laudis\Neo4j\Client\HttpDriver\Pipeline->run() #4 C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\Connection\Connection in C:\xampp\htdocs\playground CODE\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php on line 245`

The initial request 

`$client = ClientBuilder::create()
    ->addConnection('default', 'http://neo4j:test123@localhost:7474') 
    ->build();`

seems to work as I can produce an output afterwards without any error messages. But when the second request is fired 

`$client->run('MATCH (n) RETURN COUNT(n) AS number;');`

there is no output possible anymore as the error is coming at once. 

Sorry for the bad news ...
Alai

**Comment by Alaizabel** [comment]
Hello Ghlen,

just a naive thought, but is it possible to let the client create an output of the request he receives and the request he would send back after he queried the database - and then exit without serving the requesting script? In that case I could let it run here and send you the both JSON-Datasets for evaluation? 

Best,
Alai

**Comment by transistive** [comment]
Hi Alai,

Do not worry, I actually like solving these kinds of things. I cooked up a string of code to echo the request and response. Feel free to adjust the code if you want to use another method to transform the request and response to a text format:

```
$handler = \GuzzleHttp\HandlerStack::create(new \GuzzleHttp\Handler\CurlHandler());
$handler->push(\GuzzleHttp\Middleware::mapRequest(static function (\Psr\Http\Message\RequestInterface  $request) {
    echo var_dump($request);
    return $request;
}));
$handler->push(\GuzzleHttp\Middleware::mapResponse(static function (\Psr\Http\Message\ResponseInterface  $response) {
    echo var_dump($response);
    echo $response->getBody()->getContents();
    $response->getBody()->rewind();
    return $response;
}));

$httpClient = new \Http\Adapter\Guzzle6\Client(new \GuzzleHttp\Client(['handler' => $handler]));

$config = \Laudis\Neo4j\Client\HttpDriver\Configuration::create()
    ->setValue('http_client', $httpClient);

$client = \Laudis\Neo4j\Client\ClientBuilder::create()
    ->addConnection('http', 'http://neo4j:test@neo4j-new:7474', $config)
    ->build();

$client->run('MATCH (x) RETURN x');
```

I would also like to point out the way you connect over the browser, and the client is through the bolt protocol, not the HTTP one. Sadly, the new bolt protocol is not documented, and as a result is a development hell. I am hoping to find the energy and time to fix it, but that is not something for this month :sweat_smile: 

If, for some reason, this code does not work, we will try and use your strategy of the JSON Datasets.

Let's hope this output will shed some light on the problem :+1: 

Have a great weekend,
Ghlen

**Comment by Alaizabel** [comment]
Hello Ghlen,

thank you for your code. I changed the following:

1. Added in the testscript

`  require_once 'vendor/autoload.php';
  use Laudis\Neo4j\Client\ClientBuilder;`

2. Changed the 'neo4j-new' string against 'localhost' to connect in testscript, phpunit.xml.dist and docker-compose.yml
3. Changed the PW for authentication

Doing so I received the following output:

`object(GuzzleHttp\Psr7\Request)#19 (7) { ["method":"GuzzleHttp\Psr7\Request":private]=> string(3) "GET" ["requestTarget":"GuzzleHttp\Psr7\Request":private]=> NULL ["uri":"GuzzleHttp\Psr7\Request":private]=> object(GuzzleHttp\Psr7\Uri)#20 (7) { ["scheme":"GuzzleHttp\Psr7\Uri":private]=> string(4) "http" ["userInfo":"GuzzleHttp\Psr7\Uri":private]=> string(13) "neo4j:test123" ["host":"GuzzleHttp\Psr7\Uri":private]=> string(9) "localhost" ["port":"GuzzleHttp\Psr7\Uri":private]=> int(7474) ["path":"GuzzleHttp\Psr7\Uri":private]=> string(0) "" ["query":"GuzzleHttp\Psr7\Uri":private]=> string(0) "" ["fragment":"GuzzleHttp\Psr7\Uri":private]=> string(0) "" } ["headers":"GuzzleHttp\Psr7\Request":private]=> array(2) { ["User-Agent"]=> array(1) { [0]=> string(39) "GuzzleHttp/6.5.5 curl/7.70.0 PHP/7.4.10" } ["Host"]=> array(1) { [0]=> string(14) "localhost:7474" } } ["headerNames":"GuzzleHttp\Psr7\Request":private]=> array(2) { ["user-agent"]=> string(10) "User-Agent" ["host"]=> string(4) "Host" } ["protocol":"GuzzleHttp\Psr7\Request":private]=> string(3) "1.1" ["stream":"GuzzleHttp\Psr7\Request":private]=> object(GuzzleHttp\Psr7\Stream)#18 (7) { ["stream":"GuzzleHttp\Psr7\Stream":private]=> resource(79) of type (stream) ["size":"GuzzleHttp\Psr7\Stream":private]=> int(0) ["seekable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["readable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["writable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["uri":"GuzzleHttp\Psr7\Stream":private]=> string(10) "php://temp" ["customMetadata":"GuzzleHttp\Psr7\Stream":private]=> array(0) { } } } object(GuzzleHttp\Psr7\Response)#37 (6) { ["reasonPhrase":"GuzzleHttp\Psr7\Response":private]=> string(2) "OK" ["statusCode":"GuzzleHttp\Psr7\Response":private]=> int(200) ["headers":"GuzzleHttp\Psr7\Response":private]=> array(4) { ["Date"]=> array(1) { [0]=> string(29) "Sat, 19 Sep 2020 10:49:20 GMT" } ["Access-Control-Allow-Origin"]=> array(1) { [0]=> string(1) "*" } ["Content-Type"]=> array(1) { [0]=> string(30) "application/json;charset=utf-8" } ["Content-Length"]=> array(1) { [0]=> string(3) "222" } } ["headerNames":"GuzzleHttp\Psr7\Response":private]=> array(4) { ["date"]=> string(4) "Date" ["access-control-allow-origin"]=> string(27) "Access-Control-Allow-Origin" ["content-type"]=> string(12) "Content-Type" ["content-length"]=> string(14) "Content-Length" } ["protocol":"GuzzleHttp\Psr7\Response":private]=> string(3) "1.1" ["stream":"GuzzleHttp\Psr7\Response":private]=> object(GuzzleHttp\Psr7\Stream)#35 (7) { ["stream":"GuzzleHttp\Psr7\Stream":private]=> resource(85) of type (stream) ["size":"GuzzleHttp\Psr7\Stream":private]=> NULL ["seekable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["readable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["writable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["uri":"GuzzleHttp\Psr7\Stream":private]=> string(10) "php://temp" ["customMetadata":"GuzzleHttp\Psr7\Stream":private]=> array(0) { } } } { "bolt_routing" : "neo4j://localhost:7687", "transaction" : "http://localhost:7474/db/{databaseName}/tx", "bolt_direct" : "bolt://localhost:7687", "neo4j_version" : "4.1.1", "neo4j_edition" : "community" }object(GuzzleHttp\Psr7\Request)#41 (7) { ["method":"GuzzleHttp\Psr7\Request":private]=> string(4) "POST" ["requestTarget":"GuzzleHttp\Psr7\Request":private]=> NULL ["uri":"GuzzleHttp\Psr7\Request":private]=> object(GuzzleHttp\Psr7\Uri)#47 (7) { ["scheme":"GuzzleHttp\Psr7\Uri":private]=> string(4) "http" ["userInfo":"GuzzleHttp\Psr7\Uri":private]=> string(0) "" ["host":"GuzzleHttp\Psr7\Uri":private]=> string(9) "localhost" ["port":"GuzzleHttp\Psr7\Uri":private]=> int(7474) ["path":"GuzzleHttp\Psr7\Uri":private]=> string(14) "/db//tx/commit" ["query":"GuzzleHttp\Psr7\Uri":private]=> string(0) "" ["fragment":"GuzzleHttp\Psr7\Uri":private]=> string(0) "" } ["headers":"GuzzleHttp\Psr7\Request":private]=> array(6) { ["Content-Length"]=> array(1) { [0]=> string(3) "109" } ["User-Agent"]=> array(1) { [0]=> string(39) "GuzzleHttp/6.5.5 curl/7.70.0 PHP/7.4.10" } ["Host"]=> array(1) { [0]=> string(14) "localhost:7474" } ["Accept"]=> array(1) { [0]=> string(30) "application/json;charset=UTF-8" } ["Content-Type"]=> array(1) { [0]=> string(16) "application/json" } ["Authorization"]=> array(1) { [0]=> string(26) "Basic bmVvNGo6dGVzdDEyMw==" } } ["headerNames":"GuzzleHttp\Psr7\Request":private]=> array(6) { ["content-length"]=> string(14) "Content-Length" ["user-agent"]=> string(10) "User-Agent" ["host"]=> string(4) "Host" ["accept"]=> string(6) "Accept" ["content-type"]=> string(12) "Content-Type" ["authorization"]=> string(13) "Authorization" } ["protocol":"GuzzleHttp\Psr7\Request":private]=> string(3) "1.1" ["stream":"GuzzleHttp\Psr7\Request":private]=> object(GuzzleHttp\Psr7\Stream)#43 (7) { ["stream":"GuzzleHttp\Psr7\Stream":private]=> resource(103) of type (stream) ["size":"GuzzleHttp\Psr7\Stream":private]=> int(109) ["seekable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["readable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["writable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["uri":"GuzzleHttp\Psr7\Stream":private]=> string(10) "php://temp" ["customMetadata":"GuzzleHttp\Psr7\Stream":private]=> array(0) { } } } object(GuzzleHttp\Psr7\Response)#40 (6) { ["reasonPhrase":"GuzzleHttp\Psr7\Response":private]=> string(9) "Not Found" ["statusCode":"GuzzleHttp\Psr7\Response":private]=> int(404) ["headers":"GuzzleHttp\Psr7\Response":private]=> array(3) { ["Access-Control-Allow-Origin"]=> array(1) { [0]=> string(1) "*" } ["Cache-Control"]=> array(1) { [0]=> string(33) "must-revalidate,no-cache,no-store" } ["Content-Length"]=> array(1) { [0]=> string(1) "0" } } ["headerNames":"GuzzleHttp\Psr7\Response":private]=> array(3) { ["access-control-allow-origin"]=> string(27) "Access-Control-Allow-Origin" ["cache-control"]=> string(13) "Cache-Control" ["content-length"]=> string(14) "Content-Length" } ["protocol":"GuzzleHttp\Psr7\Response":private]=> string(3) "1.1" ["stream":"GuzzleHttp\Psr7\Response":private]=> object(GuzzleHttp\Psr7\Stream)#46 (7) { ["stream":"GuzzleHttp\Psr7\Stream":private]=> resource(105) of type (stream) ["size":"GuzzleHttp\Psr7\Stream":private]=> NULL ["seekable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["readable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["writable":"GuzzleHttp\Psr7\Stream":private]=> bool(true) ["uri":"GuzzleHttp\Psr7\Stream":private]=> string(10) "php://temp" ["customMetadata":"GuzzleHttp\Psr7\Stream":private]=> array(0) { } } }
Fatal error: Uncaught JsonException: Syntax error in C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php:201 Stack trace: #0 C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(201): json_decode('', true, 512, 4194304) #1 C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(270): Laudis\Neo4j\Client\HttpDriver\SessionApi4->decorateException(Object(Http\Client\Exception\HttpException)) #2 C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(92): Laudis\Neo4j\Client\HttpDriver\SessionApi4->handleRequest(Object(GuzzleHttp\Psr7\Request)) #3 C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\Pipeline.php(48): Laudis\Neo4j\Client\HttpDriver\SessionApi4->flush(Object(Laudis\Neo4j\Client\HttpDriver\Pipeline)) #4 C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php(52): Laudis\Neo4j\Client\HttpDriver\Pipeline->run() #5 C:\xa in C:\xampp\htdocs\neo_connect\vendor\laudis\neo4j-php-client\src\HttpDriver\SessionApi4.php on line 201`

Does this helps further?
Thanks,
Alai

**Comment by transistive** [comment]
Thank you for the swift response! My mistake became immediately apparent.

I made the mistake of leaving a getenv function call in the code. Neo4j v4+ requires you to specify a database. I "solved" this by adding an environment variable NEO4J_DATABASE, which determines it. If this variable does not exist, getenv returns false. False resolves to an empty string when used in the sprintf function. The connecting URI thus became /db//tx/commit. Another bug then amplified the confusion. There is a descriptive error, which raises if you use any other Uri like /db/wrongdatabasename/tx/commit. Somewhere the URL /db//tx/commit is the only one to escape this check.  :facepalm: 

So I made several changes to streamline the process:
- There is now a check for a 404 response with a descriptive error when appropriate.
- **all** getenv function calls are now removed from the source code.
- Neo4j is now the default database to connect to from v4 and up.
- There are instructions available on how to connect to a different database under the README.md file's installation section.

Sorry for the inconvenience and a huge thank you for the continuous feedback.

Please note the this is available from version 5.1.0.

Kind regards,

Ghlen

**Comment by Alaizabel** [comment]
Brilliant!
I was able to query the database now. I will close the issue and go on with some tests :)

Many thanks!
Alai
